addons/test_website/tests/test_redirect.py

```diff
@@ -231,3 +231,56 @@ def test_05_redirect_404_notfound_record(self):
         r = self.url_open(url_rec1, allow_redirects=True)
         self.assertEqual(r.status_code, 200)
         self.assertTrue(r.url.endswith(redirect.url_to))
+
+@tagged('-at_install', 'post_install')
+class TestRedirectPortal(HttpCase):
+
+    @classmethod
+    def setUpClass(cls):
+        super().setUpClass()
+        cls.user_admin = cls.env.ref('base.user_admin')
+        cls.customer = cls.env['res.partner'].create({
+            'country_id': cls.env.ref('base.fr').id,
+            'email': 'mdelvaux34@example.com',
+            'lang': 'en_US',
+            'mobile': '+33639982325',
+            'name': 'Mathias Delvaux',
+            'phone': '+33353011823',
+        })
+        cls.record_portal = cls.env['website.mail.test.portal'].create({
+            'name': 'Test Portal Record',
+            'partner_id': cls.customer.id,
+            'user_id': cls.user_admin.id,
+        })
+        for group_name, group_func, group_data in cls.record_portal.sudo()._notify_get_recipients_groups(
+            cls.env['mail.message'], False
+        ):
+            if group_name == 'portal_customer' and group_func(cls.customer):
+                cls.record_access_url = group_data['button_access']['url']
+                break
+        else:
+            raise AssertionError('Record access URL not found')
+        cls.website = cls.env.ref('website.default_website')
+
+    def test_customer_access_not_logged_language_redirection(self):
+        """Check that the customer is redirected to a page in his language."""
+        ResLang = self.env['res.lang']
+        lang_code = 'en_US'
+        ResLang._activate_lang(lang_code) or ResLang._create_lang(lang_code)
+        self.customer.lang = lang_code
+        res = self.url_open(self.record_access_url)
+        self.assertNotIn(f'/en/my/website_test_portal/{self.record_portal.id}', res.url)
+        self.assertIn(f'/my/website_test_portal/{self.record_portal.id}', res.url)
```

Comment:
Can't we add this as a lang_code / url_prefix case? So that it fits in your loop?